### PR TITLE
RFC: Selectors

### PR DESCRIPTION
**Pull request by STRRL**
Rendered markdown: [link](https://github.com/STRRL/rfcs/blob/selectors/text/2020-12-25-selectors.md)

Here is an RFC about more efficient and flexible selectors.

Feel free to make any comments.

### CODE REVIEW

**Comment by STRRL** [comment]
Hi @Hexilee, PTAL

File: text/2020-12-25-selectors.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+## Detailed design
+
+### SelectSpec
+
+The struct of CRD will not change.
+
+It will add another method for `SelectSpec`:
+
+```go
+type SelectSpec interface {
```
Comment (Hexilee) [inline text/2020-12-25-selectors.md:100]: Should I implement this interface for custom resources? How do we fetch the custom resource in k8s api? (I'm not much familiar with selector)

Comment (STRRL) [reply text/2020-12-25-selectors.md:100]: > Should I implement this interface for custom resources?

You should not implement this interface for custom resources which aims to mirrors pods.
Only Chaos CRD (like NetworkChaos, PodChaos) should implement this interface. And chaos-mesh has already done this by code generation. 

> How do we fetch the custom resource in k8s API?

Here should be an implementation for `Selector`, called `KubernetesSlector`, for example:

```go
struct KubernetesSelector {
    kubeCient client.Client
    labelSelector metav1.LabelSelector
    // other option that could apply on kubernetes native kubeclient API
}

func (it *KubernetesSelector) Select() ([]interface{}, error) {
    listOptions := client.ListOptions{}
    // fill up listOptions
    listOptions.LabelSelector = it.labelSelector
    // ... 
    var result := yourgvk.YourCRDList{}
    err := it.kubeClient.List(ctx, &result, listOptions)
    // ...
    return result
}
```

Comment (Hexilee) [reply text/2020-12-25-selectors.md:100]: I know... In this case, chaos-mesh can only list local objects like "mirrors resources". So I need to dependent on chaos-mesh api and create resources mirroring my custom resources.

Comment (STRRL) [reply text/2020-12-25-selectors.md:100]: No, you do not need Chaos Mesh API when you build your CRD.

I think your mirror resource is extremely like Pod, and obviously, Pod does not depend on Chaos Mesh.

Comment (Hexilee) [reply text/2020-12-25-selectors.md:100]: But its type should be known by chaos-mesh, shouldn't it? For example, I create a CRD `SomeResource` in package `someone`, then need we write following code in chaos-mesh:

```go
struct SomeResourceSelector {
    kubeCient client.Client
    labelSelector metav1.LabelSelector
    // other option that could apply on kubernetes native kubeclient API
}

func (it *SomeResourceSelector ) Select() ([]interface{}, error) {
    listOptions := client.ListOptions{}
    // fill up listOptions
    listOptions.LabelSelector = it.labelSelector
    // ... 
    var result := someone.SomeResourceList{}
    err := it.kubeClient.List(ctx, &result, listOptions)
    // ...
    return result
}
```

Comment (STRRL) [reply text/2020-12-25-selectors.md:100]: > But its type should be known by chaos-mesh, shouldn't it?

If you want these codes should execute in the same binary of Chaos Mesh, then YES.

Comment (Hexilee) [reply text/2020-12-25-selectors.md:100]: So, in this selector disgin, my understanding is chaos-mesh can provide a new api like `MirrorResource` following.

```go
type MirrorResourceSpec struct {
	// system which provides this mirror
	Provider strring

	// unique id, used as containerID in chaos daemon
	ID string
	
	ChaosDaemonAddress string

	// other nesessary spec
}

type MirrorResourceStatus struct {
	ExecStress struct {
		Instance string
  		StartTime int64;
	}

	// other states ...
}
```

Then other systems can create `MirrorResource` and apply chaos for their resources.

Comment (STRRL) [reply text/2020-12-25-selectors.md:100]: No, considering the complex, chaos-mesh will not provide any unnecessary API.

I think it's better to build another binary for handling customize resources as a plugin.🤔 And chaos-mesh will call this plugin with HTTP API. I think I should introduce this(how to make a plugin) in this RFC.

Comment (Hexilee) [reply text/2020-12-25-selectors.md:100]: Well, looking forward your introduction.

File: text/2020-12-25-selectors.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+struct NamespaceFilter {
+    AllowedNamespaces []string
+}
+
+func (it *NamespaceFilter) Filter(origin []interface{}) ([]interface,error) {
+    // your codes here
+}
+```
+
+### SelectorChain
```
Comment (YangKeao) [inline text/2020-12-25-selectors.md:176]: I prefer to get a single filter, which could be a combination of multiple filter. e.g. `AND` filter:

```go
AND(aFilter, bFilter, OR(cFilter, dFilter, eFilter), fFilter)
```

Comment (STRRL) [reply text/2020-12-25-selectors.md:176]: Cool!  When we accept more than one filter, we must apply them with AND/OR. `AND` and `OR` could create various filter logic with compositing many `Filter`s.

I will update this doc.

**Comment by STRRL** [comment]
I am going to close this PR because it has been too long to current context.

**Pull request closed without merging**